Sources/Core/Microsoft.StreamProcessing/Aggregates/TopKState.cs

```diff
@@ -0,0 +1,236 @@
+                else if (toRemove > min.Count)
+                {
+                    throw new InvalidOperationException("CurrentValues has more items than required");
+                }
+            }
+        }
+
+        public void Remove(T input, long timestamp)
+        {
+            if (timestamp < currentTimestamp)
+            {
+                previousValues.Remove(input);
+            }
+            else if (timestamp == currentTimestamp)
+            {
+                currentValues.Remove(input);
+            }
+            else
+            {
+                throw new ArgumentException("Invalid timestamp");
+            }
+        }
+
+        public void RemoveAll(ITopKState<T> other)
+        {
+            if (other.Count != 0)
+            {
+                if (other is HoppingTopKState<T> otherTopK)
+                {
+                    if (otherTopK.currentTimestamp >= currentTimestamp)
+                    {
+                        throw new ArgumentException("Cannot remove entries with current or future timestamp");
+                    }
+                    previousValues.RemoveAll(otherTopK.currentValues);
+                    previousValues.RemoveAll(otherTopK.previousValues);
+                }
+                else
+                {
+                    throw new InvalidOperationException("Cannot remove non-HoppingTopKState from HoppingTopKState");
+                }
+            }
+        }
+
+        public SortedMultiSet<T> GetSortedValues()
+        {
+            if (previousValues.IsEmpty)
+                return currentValues;
+            else
+            {
+                MergeCurrentToPrevious();
+                return previousValues;
+            }
+        }
+
+        private void MergeCurrentToPrevious()
+        {
+            if (!currentValues.IsEmpty)
+            {
+                // Swap so we merge small onto larger
+                if (previousValues.UniqueCount < currentValues.UniqueCount)
+                {
+                    var temp = previousValues;
+                    previousValues = currentValues;
+                    currentValues = temp;
+                }
+
+                if (!currentValues.IsEmpty)
+                {
+                    previousValues.AddAll(currentValues);
+                    currentValues.Clear();
+                }
+            }
+        }
+
+        public void AddAll(ITopKState<T> other)
+        {
+            if (other is HoppingTopKState<T> otherTopK)
+            {
+                if (otherTopK.currentTimestamp == currentTimestamp)
+                {
+                    currentValues.AddAll(otherTopK.currentValues);
+                    while (currentValues.TotalCount > k)
+                        currentValues.Remove(currentValues.First());
+                }
+                else if (otherTopK.currentTimestamp < currentTimestamp)
+                {
+                    previousValues.RemoveAll(otherTopK.currentValues);
```

Comment:
RemoveAll? Typo? If so, please make sure to add to tests to cover this, or just throw exceptions if it's not a real case.

Reply:
The AddAll is not required (it came as part of SummableAggregate which is not used). I have removed the code altogether